### ISSUE TITLE
[java] [doc] NonThreadSafeSingleton: Be more explicit as to why double checked locking is not recommended

### ISSUE THREAD

**Issue by SUPERCILEX** [pmd/pmd#216]
**Rule Set:** design.html#NonThreadSafeSingleton

**Code Sample demonstrating the issue:**

```java
if (sDatabase == null) { // <<<<<<<< incorrectly reported as not being thread safe. While this is technically true for this if statement, the double lock pattern is common enough that PMD should know about it.
    synchronized (BaseHelper.class) {
        if (sDatabase == null) {
            sDatabase = FirebaseDatabase.getInstance();
            sDatabase.setPersistenceEnabled(true);
        }
    }
}
```

**Running PMD through:** Gradle

**Comment by jsotuyod** [comment]
@SUPERCILEX thanks for your report!

Citing the docs for that rule:

> Non-thread safe singletons can result in bad state changes. Eliminate static singletons if possible by instantiating the object directly. Static singletons are usually not needed as only a single instance exists anyway. Other possible fixes are to synchronize the entire method or to use an initialize-on-demand holder class (do not use the double-check idiom).

The double checked locking pattern is broken. There is no guarantee for it to work under the Java Memory Model, *unless*, you declare the variable as `volatile`. There are plenty of good explanations on the topic, for instance http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

However, using `volatile` is not really a good fix. It carries along a performance penalty for each access (all threads MUST read the value from memory each time). I recommend using the [initialize-on-demand holder class pattern](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom), since that's the simplest solution for lazily initializing singletons in a thread-safe manner.

**Comment by SUPERCILEX** [comment]
@jsotuyod Oh, wow! Thanks for explaining all that!

So would this work?
```java
public class DatabaseHelper {
    public static DatabaseReference getReference() {
        return DatabaseHolder.INSTANCE.getReference();
    }

    private static final class DatabaseHolder {
        public static final FirebaseDatabase INSTANCE = FirebaseDatabase.getInstance();

        static {
            INSTANCE.setPersistenceEnabled(true);
        }
    }
}
```

**Comment by jsotuyod** [comment]
Yes, that seems right. I'll edit this report into a documentation issue to expand on the rule documentation / report message if that's ok with you.

**Comment by SUPERCILEX** [comment]
@jsotuyod That would be great!
